File: backend/app/schemas/facturacion.py
from datetime import date, datetime

from pydantic import BaseModel, ConfigDict, Field


# ── Detalle ────────────────────────────────────────────────────────────────────

class DetalleCreate(BaseModel):
    orden_id: int | None = None
    descripcion: str
    cantidad: int = 1
    precio_unitario: float
    subtotal: float


class DetalleRead(DetalleCreate):
    model_config = ConfigDict(from_attributes=True)
    id: int
    factura_id: int


# ── Pagos ──────────────────────────────────────────────────────────────────────

class PagoCreate(BaseModel):
    fecha_pago: date
    monto: float = Field(gt=0)
    metodo_pago: str = Field(pattern="^(efectivo|transferencia|cheque|tarjeta|otros)$")
    referencia: str | None = None
    observaciones: str | None = None


class PagoRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    factura_id: int
    fecha_pago: date
    monto: float
    metodo_pago: str
    referencia: str | None
    observaciones: str | None
    fecha_creacion: datetime | None


class PagoRealizadoRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    factura_id: int | None
    liquidacion_id: int | None
    fecha_pago: date
    monto: float
    metodo_pago: str
    referencia: str | None
    observaciones: str | None
    fecha_creacion: datetime | None


# ── Facturas emitidas ──────────────────────────────────────────────────────────

class FacturaEmitidaCreate(BaseModel):
    numero_factura: str
    cliente_id: int
    fecha_emision: date
    fecha_vencimiento: date
    periodo_mes: int = Field(ge=1, le=12)
    periodo_anio: int = Field(ge=2020, le=2030)
    cantidad_items: int = 1
    subtotal: float
    descuento: float = 0
    total: float
    observaciones: str | None = None
    detalles: list[DetalleCreate] = []
    ordenes_ids: list[int] = []      # órdenes a vincular y marcar como facturadas


class FacturaEmitidaUpdate(BaseModel):
    fecha_vencimiento: date | None = None
    subtotal: float | None = None
    descuento: float | None = None
    total: float | None = None
    saldo_pendiente: float | None = None
    estado: str | None = None
    observaciones: str | None = None


class ClienteMin(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    nombre_empresa: str
    nit: str


class FacturaEmitidaRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    numero_factura: str
    cliente_id: int
    cliente: ClienteMin
    fecha_emision: date
    fecha_vencimiento: date
    periodo_mes: int
    periodo_anio: int
    cantidad_items: int
    subtotal: float
    descuento: float
    total: float
    saldo_pendiente: float
    estado: str
    observaciones: str | None
    fecha_creacion: datetime | None
    fecha_modificacion: datetime | None
    detalles: list[DetalleRead] = []
    pagos: list[PagoRead] = []


# ── Facturas recibidas ─────────────────────────────────────────────────────────

class FacturaRecibidaCreate(BaseModel):
    numero_factura: str
    personal_id: int
    tipo: str = Field(pattern="^(courier|transportadora|materiales|otros)$")
    fecha_recepcion: date
    fecha_vencimiento: date
    periodo_mes: int | None = None
    periodo_anio: int | None = None
    cantidad_items: int = 0
    subtotal: float
    descuento: float = 0
    total: float
    observaciones: str | None = None
    detalles: list[DetalleCreate] = []


class FacturaRecibidaUpdate(BaseModel):
    fecha_vencimiento: date | None = None
    subtotal: float | None = None
    total: float | None = None
    saldo_pendiente: float | None = None
    estado: str | None = None
    observaciones: str | None = None


class PersonalMin(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    codigo: str
    nombre_completo: str
    tipo_personal: str


class FacturaRecibidaRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    numero_factura: str
    personal_id: int
    personal: PersonalMin
    tipo: str
    fecha_recepcion: date
    fecha_vencimiento: date
    periodo_mes: int | None
    periodo_anio: int | None
    cantidad_items: int
    subtotal: float
    descuento: float
    total: float
    saldo_pendiente: float
    estado: str
    observaciones: str | None
    fecha_creacion: datetime | None
    fecha_modificacion: datetime | None
    detalles: list[DetalleRead] = []
    pagos: list[PagoRealizadoRead] = []


# ── Cuentas y resumen ──────────────────────────────────────────────────────────

class CuentaItem(BaseModel):
    id: int
    tipo: str
    referencia: str
    codigo: str | None
    acreedor_o_deudor: str
    fecha_vencimiento: date
    monto: float
    estado: str
    dias: int
    clasificacion: str


class ResumenFinanciero(BaseModel):
    total_por_cobrar: float
    total_vencido_cobrar: float
    vence_esta_semana_cobrar: float
    total_por_pagar: float
    total_vencido_pagar: float
    vence_esta_semana_pagar: float
    facturas_emitidas_mes: float
    facturas_recibidas_mes: float

File: frontend/src/api/facturacion.ts
import api from "./client";

export interface DetalleFactura {
  id?: number;
  orden_id?: number | null;
  descripcion: string;
  cantidad: number;
  precio_unitario: number;
  subtotal: number;
}

export interface Pago {
  id: number;
  factura_id: number;
  fecha_pago: string;
  monto: number;
  metodo_pago: string;
  referencia: string | null;
  observaciones: string | null;
  fecha_creacion: string | null;
}

export interface FacturaEmitida {
  id: number;
  numero_factura: string;
  cliente_id: number;
  cliente: { id: number; nombre_empresa: string; nit: string };
  fecha_emision: string;
  fecha_vencimiento: string;
  periodo_mes: number;
  periodo_anio: number;
  cantidad_items: number;
  subtotal: number;
  descuento: number;
  total: number;
  saldo_pendiente: number;
  estado: "pendiente" | "parcial" | "pagada" | "vencida" | "anulada";
  observaciones: string | null;
  fecha_creacion: string | null;
  detalles: DetalleFactura[];
  pagos: Pago[];
}

export interface FacturaRecibida {
  id: number;
  numero_factura: string;
  personal_id: number;
  personal: { id: number; codigo: string; nombre_completo: string; tipo_personal: string };
  tipo: "courier" | "transportadora" | "materiales" | "otros";
  fecha_recepcion: string;
  fecha_vencimiento: string;
  periodo_mes: number | null;
  periodo_anio: number | null;
  cantidad_items: number;
  subtotal: number;
  descuento: number;
  total: number;
  saldo_pendiente: number;
  estado: "pendiente" | "parcial" | "pagada" | "anulada";
  observaciones: string | null;
  detalles: DetalleFactura[];
  pagos: Pago[];
}

export interface ResumenFinanciero {
  total_por_cobrar: number;
  total_vencido_cobrar: number;
  vence_esta_semana_cobrar: number;
  total_por_pagar: number;
  total_vencido_pagar: number;
  vence_esta_semana_pagar: number;
  facturas_emitidas_mes: number;
  facturas_recibidas_mes: number;
}

export interface CuentaItem {
  id: number;
  tipo: string;
  referencia: string;
  codigo: string | null;
  acreedor_o_deudor: string;
  fecha_vencimiento: string;
  monto: number;
  estado: string;
  dias: number;
  clasificacion: string;
}

export const facturacionApi = {
  // Resumen y cuentas
  resumen: () => api.get<ResumenFinanciero>("/facturacion/resumen"),
  cuentasPorCobrar: () => api.get<CuentaItem[]>("/facturacion/cuentas-por-cobrar"),
  cuentasPorPagar: () => api.get<CuentaItem[]>("/facturacion/cuentas-por-pagar"),

  // Emitidas
  listEmitidas: (params?: { cliente_id?: number; estado?: string; periodo_mes?: number; periodo_anio?: number }) =>
    api.get<FacturaEmitida[]>("/facturacion/emitidas", { params }),
  getEmitida: (id: number) => api.get<FacturaEmitida>(`/facturacion/emitidas/${id}`),
  createEmitida: (data: Partial<FacturaEmitida> & { ordenes_ids?: number[] }) =>
    api.post<FacturaEmitida>("/facturacion/emitidas", data),
  updateEmitida: (id: number, data: Partial<FacturaEmitida>) =>
    api.put<FacturaEmitida>(`/facturacion/emitidas/${id}`, data),
  anularEmitida: (id: number) => api.delete(`/facturacion/emitidas/${id}`),
  registrarPagoEmitida: (id: number, pago: Omit<Pago, "id" | "factura_id" | "fecha_creacion">) =>
    api.post<Pago>(`/facturacion/emitidas/${id}/pagos`, pago),

  // Recibidas
  listRecibidas: (params?: { personal_id?: number; tipo?: string; estado?: string }) =>
    api.get<FacturaRecibida[]>("/facturacion/recibidas", { params }),
  getRecibida: (id: number) => api.get<FacturaRecibida>(`/facturacion/recibidas/${id}`),
  createRecibida: (data: Partial<FacturaRecibida>) =>
    api.post<FacturaRecibida>("/facturacion/recibidas", data),
  updateRecibida: (id: number, data: Partial<FacturaRecibida>) =>
    api.put<FacturaRecibida>(`/facturacion/recibidas/${id}`, data),
  registrarPagoRecibida: (id: number, pago: Omit<Pago, "id" | "factura_id" | "fecha_creacion">) =>
    api.post<Pago>(`/facturacion/recibidas/${id}/pagos`, pago),
};
